crates/iceberg/src/input.rs

```diff
@@ -0,0 +1,886 @@
+        &self,
+        input_stream: &mut dyn ArrowStream,
+        schema: &Relation,
+        receiver: &mut Receiver<PipelineState>,
+    ) {
+        self.read_ordered_snapshot_inner(input_stream, schema, receiver)
+            .await
+            .unwrap_or_else(|e| self.consumer.error(true, e));
+    }
+
+    async fn read_ordered_snapshot_inner(
+        &self,
+        input_stream: &mut dyn ArrowStream,
+        schema: &Relation,
+        receiver: &mut Receiver<PipelineState>,
+    ) -> Result<(), AnyError> {
+        let timestamp_column = self.config.timestamp_column.as_ref().unwrap();
+
+        let timestamp_field = schema.field(timestamp_column).unwrap();
+
+        // The following unwraps are safe, as validated in `validate_timestamp_column`.
+        let lateness = timestamp_field.lateness.as_ref().unwrap();
+
+        // Query the table for min and max values of the timestamp column that satisfy the filter.
+        let bounds_query =
+            format!("select * from (select cast(min({timestamp_column}) as string) as start_ts, cast(max({timestamp_column}) as string) as end_ts from snapshot {}) where start_ts is not null",
+            if let Some(filter) = &self.config.snapshot_filter {
+                format!("where {filter}")
+            } else {
+                String::new()
+            });
+
+        let bounds = execute_query_collect(&self.datafusion, &bounds_query).await?;
+
+        info!(
+            "iceberg {}: querying the table for min and max timestamp values",
+            &self.endpoint_name,
+        );
+
+        if bounds.len() != 1 || bounds[0].num_rows() != 1 {
+            info!(
+                "iceberg {}: initial snapshot is empty; the Delta table contains no records{}",
+                &self.endpoint_name,
+                if let Some(filter) = &self.config.snapshot_filter {
+                    format!(" that satisfy the filter condition '{filter}'")
+                } else {
+                    String::new()
+                }
+            );
+            return Ok(());
+        }
+
+        if bounds[0].num_columns() != 2 {
+            // Should never happen.
+            return Err(anyhow!(
+                    "internal error: query '{bounds_query}' returned a result with {} columns; expected 2 columns",
+                    bounds[0].num_columns()
+                ));
+        }
+
+        let min = bounds[0]
+            .column(0)
+            .as_string_opt::<i32>()
+            .ok_or_else(|| anyhow!("internal error: cannot retrieve the output of query '{bounds_query}' as a string"))?
+            .value(0)
+            .to_string();
+
+        let max = bounds[0].column(1).as_string::<i32>().value(0).to_string();
+
+        info!(
+            "iceberg {}: reading table snapshot in the range '{min} <= {timestamp_column} <= {max}'",
+            &self.endpoint_name,
+        );
+
+        let min = timestamp_to_sql_expression(&timestamp_field.columntype, &min);
```

Comment:
this part looks familiar, couldn't put it into a library?

Reply:
I moved some code to a library, but I expect this code may evolve separately in iceberg and delta